### PR TITLE
refactor(osmomath): limit pow iterations in osmomath

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

Limits the number of Pow iterations:
To make sure that we don't hit the loop iteration bound too often
under normal conditions, this change was backported to v19.x (mainnet at the time of writing)
and tested for no app-hash over 10000 blocks.

```
go test -benchmem -run=^$ -bench ^BenchmarkSlowPowCases$ github.com/osmosis-labs/osmosis/osmomath -count=6
goos: linux
goarch: amd64
pkg: github.com/osmosis-labs/osmosis/osmomath
cpu: 12th Gen Intel(R) Core(TM) i7-1260P
BenchmarkSlowPowCases-16              76          14897602 ns/op         3760805 B/op     168182 allocs/op
BenchmarkSlowPowCases-16              78          14858077 ns/op         3760792 B/op     168181 allocs/op
BenchmarkSlowPowCases-16              79          15163556 ns/op         3760784 B/op     168181 allocs/op
BenchmarkSlowPowCases-16              79          15022864 ns/op         3760810 B/op     168182 allocs/op
BenchmarkSlowPowCases-16              85          15068015 ns/op         3760784 B/op     168182 allocs/op
BenchmarkSlowPowCases-16              68          15159170 ns/op         3760789 B/op     168182 allocs/op
PASS
ok      github.com/osmosis-labs/osmosis/osmomath        7.105s
```

## Testing and Verifying

- This change is a trivial rework / code cleanup without any test coverage.
- Was tested by backporting to the v19.x release line

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: osmomath/math.go [hunk 1]
```diff
@@ -8,6 +8,8 @@ import (
 // TODO: Analyze choice here.
 var powPrecision, _ = NewDecFromStr("0.00000001")
 
+const powIterationLimit = 800_000
```
Comment (pysel) [inline osmomath/math.go:11]: asking to learn: is there a specific reason for this number?

File: x/gamm/pool-models/balancer/pool_suite_test.go [hunk 2]
```diff
@@ -325,6 +325,36 @@ var calcSingleAssetJoinTestCases = []calcJoinSharesTestCase{
...
+		//
+		// where:
+		// 	P_supply = initial pool supply = 1e20
+		//	A_t = amount of deposited asset = 195920
+		//	B_t = existing balance of deposited asset in the pool prior to deposit = 156_736
+		//	W_t = normalized weight of deposited asset in pool = 100 / (100 + 1000) approx = 0.09
+		// Plugging all of this in, we get:
+		// 	Full solution: https://www.wolframalpha.com/input?i=100+*10%5E18*%28%281+%2B++%28499990*%281+-+%281-%28100+%2F+%28100++%2B+1000%29%29%29+*+0%29%2F500000%29%29%5E%28100+%2F+%28100+%2B++1000%29%29+-+1%29
+		// 	Simplified:  P_issued = 6_504_012_121_638_943_579
+		name:         "11single asset - (almost 1 == tokenIn / liquidity ratio), token in weight is smaller than the other token, with zero spread factor",
```
Comment (p0mvn) [inline x/gamm/pool-models/balancer/pool_suite_test.go:342]:
```suggestion
		name:         "single asset - (almost 1 == tokenIn / liquidity ratio), token in weight is smaller than the other token, with zero spread factor",
```

File: x/gamm/pool-models/balancer/pool_suite_test.go [hunk 3]
```diff
@@ -325,6 +325,36 @@ var calcSingleAssetJoinTestCases = []calcJoinSharesTestCase{
...
+				Weight: osmomath.NewInt(100),
+			},
+			{
+				Token:  sdk.NewInt64Coin("uatom", 1e12),
+				Weight: osmomath.NewInt(1000),
+			},
+		},
+		// Increasing this by 1 would cause a panic due to pow iteration limit reached
+		tokensIn:     sdk.NewCoins(sdk.NewInt64Coin("uosmo", 499_990)),
+		expectShares: osmomath.NewIntFromUint64(6_504_012_121_638_943_579),
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/gamm/pool-models/balancer/pool_suite_test.go]: The original tokenIn value now causes an iteration limit. The test covering that behavior is kept.

The new token in value of `499990` is the closest one to the old that does not panic

**Review by czarcas7ic — APPROVED** [review]

File: osmomath/math.go [hunk 4]
```diff
@@ -8,6 +8,8 @@ import (
 // TODO: Analyze choice here.
 var powPrecision, _ = NewDecFromStr("0.00000001")
 
+const powIterationLimit = 150_000
```
Comment (mattverse) [inline osmomath/math.go:11]: Curious, how was this chosen? 👀

**Review by mattverse — COMMENTED** [review]
The change lgtm, we didn't include this in a release yet right?
I feel like this change should definitely be state breaking thus should be included in the next major version upgrade

File: osmomath/math.go (hunk 4 above)
Comment (p0mvn) [reply osmomath/math.go:11]: https://github.com/osmosis-labs/osmosis/pull/6627#discussion_r1345752782

**Comment by p0mvn** [comment]
> The change lgtm, we didn't include this in a release yet right? I feel like this change should definitely be state breaking thus should be included in the next major version upgrade

This is v20.

**Pull request merged**